programs/mango-v4/src/instructions/openbook_v2_place_order.rs

```diff
@@ -0,0 +1,565 @@
+        );
+
+        // Validate bank and vault #3
+        let payer_bank = ctx.accounts.bank.load()?;
+        require_keys_eq!(payer_bank.vault, ctx.accounts.vault.key());
+        let payer_token_index = match order.side {
+            OpenbookV2Side::Bid => openbook_market.quote_token_index,
+            OpenbookV2Side::Ask => openbook_market.base_token_index,
+        };
+        require_eq!(payer_bank.token_index, payer_token_index);
+
+        receiver_token_index = match order.side {
+            OpenbookV2Side::Bid => openbook_market.base_token_index,
+            OpenbookV2Side::Ask => openbook_market.quote_token_index,
+        };
+    }
+
+    //
+    // Pre-health computation
+    //
+    let mut account = ctx.accounts.account.load_full_mut()?;
+    let retriever = new_fixed_order_account_retriever(ctx.remaining_accounts, &account.borrow())?;
+    let now_ts: u64 = Clock::get()?.unix_timestamp.try_into().unwrap();
+    let mut health_cache = new_health_cache(&account.borrow(), &retriever, now_ts)
+        .context("pre-withdraw init health")?;
+    let pre_health_opt = if !account.fixed.is_in_health_region() {
+        let pre_init_health = account.check_health_pre(&health_cache)?;
+        Some(pre_init_health)
+    } else {
+        None
+    };
+
+    // Check if the bank for the token whose balance is increased is in reduce-only mode
+    let receiver_bank_ai;
+    let receiver_bank_oracle;
+    let receiver_bank_reduce_only;
+    {
+        // The token position already exists, but we need the active_index.
+        let (_, _, active_index) = account.ensure_token_position(receiver_token_index)?;
+        let group_key = ctx.accounts.group.key();
+        let (receiver_bank, oracle) =
+            retriever.bank_and_oracle(&group_key, active_index, receiver_token_index)?;
+        receiver_bank_oracle = oracle;
+        receiver_bank_reduce_only = receiver_bank.are_deposits_reduce_only();
+
+        // The fixed_order account retriever can't give us mut references, so use the above
+        // call to .bank_and_oracle() as validation and then copy out the matching AccountInfo.
+        receiver_bank_ai = ctx.remaining_accounts[active_index].clone();
+        // Double-check that we got the right account
+        let receiver_bank2 = receiver_bank_ai.load::<Bank>()?;
+        assert_eq!(receiver_bank2.group, group_key);
+        assert_eq!(receiver_bank2.token_index, receiver_token_index);
+    }
+
+    drop(retriever);
+
+    // No version check required, bank writable from v1
+
+    //
+    // Before-order tracking
+    //
+
+    let before_vault = ctx.accounts.vault.amount;
+
+    let before_oo_free_slots;
+    let before_had_bids;
+    let before_had_asks;
+    let before_oo = {
+        let open_orders = ctx.accounts.open_orders.load()?;
+        before_oo_free_slots = MAX_OPEN_ORDERS
+            - open_orders
+                .all_orders_in_use()
+                .collect::<Vec<&OpenOrder>>()
+                .len();
```

Comment:
skip the .collect() and just use .count() on the iterator for a cu and memory reduction (same below)